xwayland/selection/selection.c

```diff
@@ -33,15 +46,48 @@ void xwm_selection_transfer_destroy_property_reply(
 }
 
 void xwm_selection_transfer_init(struct wlr_xwm_selection_transfer *transfer) {
+	struct wlr_xwm *xwm = transfer->selection->xwm;
+
 	transfer->wl_client_fd = -1;
+
+	transfer->window = xcb_generate_id(xwm->xcb_conn);
+	xcb_create_window(xwm->xcb_conn,
+		XCB_COPY_FROM_PARENT,
+		transfer->window,
+		xwm->screen->root,
+		0, 0,
+		10, 10,
+		0,
+		XCB_WINDOW_CLASS_INPUT_OUTPUT,
+		xwm->screen->root_visual,
+		XCB_CW_EVENT_MASK, (uint32_t[]){
+			XCB_EVENT_MASK_SUBSTRUCTURE_NOTIFY | XCB_EVENT_MASK_PROPERTY_CHANGE
+		}
+	);
+
+	xcb_set_selection_owner(xwm->xcb_conn, transfer->window,
+			xwm->atoms[CLIPBOARD_MANAGER], XCB_TIME_CURRENT_TIME);
+	xcb_flush(xwm->xcb_conn);
 }
 
 void xwm_selection_transfer_finish(
 		struct wlr_xwm_selection_transfer *transfer) {
+	if (!transfer) {
+		return;
+	}
+
 	transfer->incr = false;
 	xwm_selection_transfer_destroy_property_reply(transfer);
 	xwm_selection_transfer_remove_event_source(transfer);
 	xwm_selection_transfer_close_wl_client_fd(transfer);
+
+	if (transfer->window) {
+		xcb_destroy_window(transfer->selection->xwm->xcb_conn, transfer->window);
+		transfer->window = XCB_WINDOW_NONE;
```

Comment:
Nit: no need to reset the window, we'll free the whole struct right after

Reply:
Done.